independent-projects/proxy/src/main/java/io/quarkus/proxy/server/ProxyServer.java

```diff
@@ -0,0 +1,121 @@
+     *                 Handler's class should be annotated with {@link ChannelHandler.Sharable}
+     *                 as it is shared among all channels and therefore invoked in multiple event loop
+     *                 group threads.
+     * @return a {@link Future} that completes with the outcome of establishing a listening socket
+     */
+    public Future<Channel> listen(
+        String host,
+        int port,
+        ChannelHandler... handlers
+    ) {
+        var eventLoop = context.nettyEventLoop();
+        var vertx = context.owner();
+        var acceptorGroup = vertx.getAcceptorEventLoopGroup();
+
+        var bootstrap = new ServerBootstrap();
+        bootstrap.channel(NioServerSocketChannel.class);
+        bootstrap.group(acceptorGroup, eventLoop);
+
+        bootstrap.childHandler(new ChannelInitializer<>() {
+            @Override
+            protected void initChannel(Channel ch) {
+                ChannelPipeline pipeline = ch.pipeline();
+
+                if (Ssl.isEnabled(ssl)) {
+                    try (
+                        var keyCertChainStream = ssl.keyCertChainUrl().get().openStream();
+                        var keyStream = ssl.keyUrl().get().openStream()
+                    ) {
+                        var sslBuilder = SslContextBuilder
+                            .forServer(keyCertChainStream, keyStream);
+                        if (ssl.trustCertCollectionUrl().isPresent()) {
+                            try (var trustCertCollectionStream = ssl.trustCertCollectionUrl().get().openStream()) {
+                                sslBuilder.trustManager(trustCertCollectionStream)
+                                          .clientAuth(ssl.clientAuth().orElse(ClientAuth.REQUIRE));
+                            }
+                        } else {
+                            sslBuilder.clientAuth(ssl.clientAuth().orElse(ClientAuth.NONE));
+                        }
+                        pipeline.addLast(sslBuilder.build().newHandler(ch.alloc()));
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }
+
+                pipeline.addLast(handlers);
+
+                if (log.isDebugEnabled()) {
+                    pipeline.addLast(new LoggingChannelInboundHandlerAdapter(log));
+                }
+            }
+        });
+
+        // Bind the server socket
+        var channelFuture = bootstrap.bind(host, port);
+
+        return VertxUtil
+            .toPromise(context, channelFuture)
+            .future()
+            .map(v -> channelFuture.channel())
+            .onSuccess(ch -> {
+                log.info("Proxy server listening on {} using {}",
+                         ch.localAddress(), Ssl.isEnabled(ssl) ? "SSL" : "TCP");
+            })
+            .onFailure(ex -> {
+                log.error("Proxy server can't establish listening channel", ex);
+            });
```

Comment:
Let's try and limit the use of lambdas and streams as much as possible

Reply:
Why?
This is simple Future mapping / handling ...

Reply:
We shy away from them as much possible for runtime code for runtime footprint reasons